x/valset-pref/validator_set.go

```diff
@@ -223,6 +226,85 @@ func (k Keeper) UndelegateFromValidatorSet(ctx sdk.Context, delegatorAddr string
+// The method then computes the total amount delegated and the amount to undelegate for each validator under this
+// newly calculated valset-ratio set.
+//
+// If the undelegation amount is greater than the total delegated amount, it returns an error.
+// The validators are then sorted in descending order of VRatio.
+// The method ensures that the largest VRatio is under 1. If it is greater than 1, it returns an error.
+// Finally, the method undelegates the target amount from each validator.
+// If an error occurs during the undelegation process, it is returned.
+func (k Keeper) UndelegateFromRebalancedValidatorSet(ctx sdk.Context, delegatorAddr string, undelegation sdk.Coin) error {
+	// GetValSetPreferencesWithDelegations fetches the delegator's validator set preferences, but returns a set with
+	// modified weights that consider their existing delegations. If there is no existing delegation, it returns an error.
+	// The new weights based on the existing delegations is returned, but the original valset preferences
+	// are not modified.
+	// For example, if someone's valset is 50/50 between two validators, but they have 10 OSMO delegated to validator A,
+	// and 90 OSMO delegated to validator B, the returned valset preference weight will be 10/90.
+
+	existingSet, err := k.GetValSetPreferencesWithDelegations(ctx, delegatorAddr)
+	if err != nil {
+		return types.NoValidatorSetOrExistingDelegationsError{DelegatorAddr: delegatorAddr}
+	}
+
+	delegator := sdk.MustAccAddressFromBech32(delegatorAddr)
+
+	// Step 1,2: compute the total amount delegated and the amount to undelegate for each validator
+	// under valset-ratios.
+	valSetRatio, validators, totalDelegatedAmt, err := k.getValsetRatios(ctx, delegator, existingSet.Preferences, undelegation.Amount)
+	if err != nil {
+		return err
+	}
+
+	if undelegation.Amount.ToLegacyDec().GT(totalDelegatedAmt) {
+		return types.UndelegateMoreThanDelegatedError{TotalDelegatedAmt: totalDelegatedAmt, UndelegationAmt: undelegation.Amount}
+	}
+
+	// Step 3: Sort validators in descending order of VRatio.
+	sort.Slice(valSetRatio, func(i, j int) bool {
+		return valSetRatio[i].VRatio.GT(valSetRatio[j].VRatio)
+	})
+
+	totalUnDelAmt := sdk.NewInt(0)
+	var amountToUnDelegate osmomath.Int
+
+	// Ensure largest VRatio is under 1.
+	// Since we called GetValSetPreferencesWithDelegations, there should be no VRatio > 1
+	if valSetRatio[0].VRatio.GT(sdk.OneDec()) {
+		return types.ValsetRatioGreaterThanOneError{ValsetRatio: valSetRatio[0].VRatio}
+	}
+
+	// Step 4: Undelegate target amount from each validator
+	for index, val := range valSetRatio {
+		validator := validators[val.ValAddr.String()]
+
+		// in the last valset iteration we don't calculate it from shares using decimals and trucation,
+		// we use whats remaining to get more accurate value
+		if len(existingSet.Preferences)-1 == index {
+			amountToUnDelegate = undelegation.Amount.Sub(totalUnDelAmt).ToLegacyDec().TruncateInt()
+		} else {
```

Comment:
Lets make amountToUndelegate = Min(amountToUndelegate, amount_staked_to_this_val)

just in case some rounding issue increases it.

Reply:
Added the logic you described https://github.com/osmosis-labs/osmosis/pull/6685/commits/cc584e704b8a83c804acc26d0d4fbb16b27d7c09

Although I am not quite sure how to force this in a gotest, seems like just a guard rail.